File: src/com/ziggybadans/keynumbergenerator/KeyNumberGenerator.java
package com.ziggybadans.keynumbergenerator;

import java.awt.Toolkit;
import java.awt.datatransfer.Clipboard;
import java.awt.datatransfer.StringSelection;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Locale;
import java.util.Objects;

/**
 * The KeyNumberGenerator class is the logic behind the program.
 * It stores hard-coded lists for the markets, durations, and types.
 * It handles the getting and setting of each variable.
 * It generates the key number and creates the GUI.
 */
public class KeyNumberGenerator {
    /**
     * Each mutable value has a variable to store the user's input and a boolean to check if the variable is valid and ready to be a part of the key number. If not, it is false.
     */

    public int year = -1;
    public boolean yearReady;

    /**
     * The inputs that are interacted through JComboBoxes use arrays to hard-code their values.
     */
    public static String[] markets = {"AUG", "BAL", "BEG", "BEN", "BER", "BOW", "BUN", "BUR", "CAI", "DAR", "DEV", "GEE", "GLA", "GOL", "HOB", "IPS", "LAU", "LIN", "MAC",
            "MAR", "MIL", "MUR", "MUS", "NOW", "PIR", "QUE", "ROC", "TOW", "WOL", "TRSN", "AGEN"};
    public boolean marketReady = false;
    public String market;

    public char writerInitial;
    public String clientInitial;

    public static Integer[] durations = {10, 15, 30, 45, 60, 90};
    public boolean durationReady = false;
    public int duration = -1;

    public static String[] types = {"R", "L", "SL", "X", "SX"};
    public boolean typeReady = false;
    public String type;

    /**
     * This number integer goes up by one every time a key number is generated, and is stored in properties by default.
     */
    public static int number = 0;
    String keyNumber;
    boolean debug = true;

    KeyNumberGenerator() {
    }

    /**
     * This method sets the market to the user's input from the {@code GUI}.
     * @param input the string to check against the {@code markets} array. If there is a match, it sets that value.
     */
    public void setMarket(String input) {
        if (input.equals("NULL")) {
            market = null;
            marketReady = false;
        } else {
            int tempMarket;
            try {
                tempMarket = java.util.Arrays.asList(markets).indexOf(input);
                market = markets[tempMarket];
                marketReady = true;
                if (debug) { System.out.println("Debug: " + market + " - Successful!"); }
            }
            catch (Exception e) {
                if (debug) { System.out.println("Debug: " + input + " - Invalid."); }
            }
        }
    }

    /**
     * This method sets {@code year} to the user's input from the {@code GUI}. It updates the minimum year by {@code Calender}.
     * A user can enter either the full year (eg. 2021) or a shortened version (eg. 21) and get the value suitable for the key number (eg. 21).
     * @param input the string to check against the current year. If there is a match, it ensures it is in the correct format, and sets that value. If there is a null value for {@code input}, it sets the year to {@code -1} and {@code yearReady} to {@code false}.
     * @see java.util.Calendar
     */
    public void setYear(String input) {
        CharConcatenation charConcat = new CharConcatenation();
        int currentYear = Calendar.getInstance().get(Calendar.YEAR);
        System.out.println("Current year: " + currentYear);
        try {
            if (input.length() == 2 && Integer.parseInt(input) >= Integer.parseInt(Integer.toString(currentYear).substring(2,4))) {
                this.year = Integer.parseInt(input);
                this.yearReady = true;
                if (debug) { System.out.println("Debug1: " + this.year + " - Successful!"); }
            } else if (input.length() == 4 && Integer.parseInt(input) >= currentYear) {
                char firstYearChar = input.charAt(2);
                char secondYearChar = input.charAt(3);
                this.year = Integer.parseInt(charConcat.concat(firstYearChar, secondYearChar));
                this.yearReady = true;
                if (debug) { System.out.println("Debug2: " + this.year + " - Successful!"); }
            } else {
                this.yearReady = false;
                if (debug) { System.out.println("Debug: " + input + " - Invalid."); }
            }
        } catch (NullPointerException e) {
            if (debug) { System.out.println("Input is null! Resetting..."); }
            year = -1;
            this.yearReady = false;
        }
    }

    /**
     * This method sets {@code writerInitial} to the user's input from the {@code GUI}.
     * A user can either enter their full name (eg. Ziggy) or just the initial (eg. Z) and get the value suitable for the key number (eg. Z).
     * @param input the string to convert to an initial. If the value is null, it sets {@code writerInitial} to an escaped null value.
     */
    public void setWriterI(String input) {
        try {
            writerInitial = input.toUpperCase(Locale.ROOT).charAt(0);
            if (debug) { System.out.println("Debug: " + writerInitial + " - Successful!"); }
        } catch (NullPointerException e) {
            if (debug) { System.out.println("Input is null! Resetting..."); }
            writerInitial = '\0';
        }
    }

    /**
     * This method sets {@code duration} to the user's input from the {@code GUI}.
     * @param input the integer to check against the {@code durations} array. If there is a match, it sets that value.
     */
    public void setDuration(int input) {
        int tempDuration;
        if (input == 0) {
            duration = 0;
            durationReady = false;
        } else {
            try {
                tempDuration = Arrays.asList(durations).indexOf(input);
                duration = durations[tempDuration];
                durationReady = true;
                if (debug) { System.out.println("Debug: " + duration + " - Successful!"); }
            }
            catch (Exception e) {
                if (debug) { System.out.println("Debug: " + input + " - Invalid."); }
                durationReady = false;
            }
        }
    }

    /**
     * This method sets {@code type} to the user's input from the {@code GUI}.
     * @param input the string to check against the {@code types} array. If there is a match, it sets that value.
     */
    public void setType(String input) {
        int tempType;
        if (input.equals("NULL")) {
            type = null;
            typeReady = false;
        }
        try {
            tempType = java.util.Arrays.asList(types).indexOf(input);
            type = types[tempType];
            typeReady = true;
            if (debug) { System.out.println("Debug " + type + " - Successful!"); }
        }
        catch (Exception e) {
            if (debug) { System.out.println("Debug: " + input + " - Invalid."); }
        }
    }

    /**
     * This method sets {@code clientInitial} to the user's input from the {@code GUI}.
     * A user can either enter the full client name (eg. Ziggy Badans) or just the initials (eg. ZB) and get the value suitable for the key number (eg. ZB).
     * @param input the string to convert to two initials. If the value is null, it sets {@code clientInitial} to null.
     */
    public void setClientI(String input) {
        try {
            CharConcatenation charConcat = new CharConcatenation();
            if (input.length() == 2) {
                char firstClientInitial = input.toUpperCase(Locale.ROOT).charAt(0);
                char secondClientInitial = input.toUpperCase(Locale.ROOT).charAt(1);

                clientInitial = charConcat.concat(firstClientInitial, secondClientInitial);
            } else {
                int secondInitialIndex = input.toUpperCase(Locale.ROOT).indexOf(" ") + 1;
                char firstClientInitial = input.toUpperCase(Locale.ROOT).charAt(0);
                char secondClientInitial = input.toUpperCase(Locale.ROOT).charAt(secondInitialIndex);

                clientInitial = charConcat.concat(firstClientInitial, secondClientInitial);
            }
            if (debug) { System.out.println("Debug: " + clientInitial + " - Successful!"); }
        } catch (NullPointerException e) {
            if (debug) { System.out.println("Input is null! Resetting..."); }
            clientInitial = null;
        }
    }

    /**
     * These methods check if the value of the associated variables are null.
     * @param input String, integer, or char to check for a null value.
     * @return a String that has the text "NULL" for the key number.
     */
    public String get(String input) {
        return Objects.requireNonNullElse(input, "NULL");
    }

    public String get(int input) {
        if (input != -1) {
            return String.valueOf(input);
        } else {
            return "NULL";
        }
    }

    public String get(char input) {
        if (input != '\0') {
            return String.valueOf(input);
        } else {
            return "NULL";
        }
    }

    /**
     * This method generates the key number. An example key number looks like this: AUG21G-ZB-10R-1. After generating, it also copies the {@code keyNumber} automatically to the clipboard.
     * @return a String that is the key number.
     */
    public String generate() {
        System.out.println("Key Number:");
        keyNumber = get(market) + get(year) + get(writerInitial) + '-' + get(clientInitial) + '-' + get(duration) + get(type) + '-' + number;
        System.out.println(keyNumber);
        number++;

        StringSelection selection = new StringSelection(keyNumber);
        Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
        clipboard.setContents(selection, selection);

        return keyNumber;
    }

    public static void main(String[] args) throws InterruptedException {
        new GUI();
    }
}

File: KeyNumberGenerator.cs
using System;
using System.Text;

namespace KeyNumberGenerator
{
    class KeyNumberGenerator
    {
        int year = -1;
        public bool yearReady;

        public static string[] markets = { "AUG", "BAL", "BEG", "BEN", "BER", "BOW", "BUN", "BUR", "CAI", "DAR", "DEV", "GEE", "GLA", "GOL", "HOB", "IPS", "LAU", "LIN", "MAC",
            "MAR", "MIL", "MUR", "MUS", "NOW", "PIR", "QUE", "ROC", "TOW", "WOL", "TRSN", "AGEN" };
        public bool marketReady = false;
        string market;

        string writerInitial;
        string clientInitial;

        public static int[] durations = { 10, 15, 30, 45, 60, 90 };
        public bool durationReady = false;
        int duration = -1;

        public static string[] types = { "R", "L", "SL", "X", "SX" };
        public bool typeReady = false;
        string type;

        static int number = 0;
        public string keyNumber;

        public KeyNumberGenerator()
        {
            number = Properties.Settings.Default.number;
        }

        public void SetMarket(string input)
        {
            if (input == "NULL")
            {
                this.market = null;
                this.marketReady = false;
                Console.WriteLine("SetMarket" + market);
            }
            else
            {
                try
                {
                    int checkInput = Array.IndexOf(markets, input);
                    this.market = markets[checkInput];
                    this.marketReady = true;
                    Console.WriteLine("SetMarket: " + market + " - Successful!");
                }
                catch (IndexOutOfRangeException)
                {
                    Console.WriteLine("SetMarket: " + input + " - Invalid.");
                }
            }
        }

        public void SetYear(String input)
        {
            string currentYear = DateTime.Now.Year.ToString();
            Console.WriteLine("Current year: " + currentYear);
            try
            {
                if (input.Length == 2 && int.Parse(input) >= int.Parse(CharConcatenation.Concat(currentYear[2], currentYear[3])))
                {
                    this.year = int.Parse(input);
                    this.yearReady = true;
                    Console.WriteLine("SetYear1: " + this.year + " - Successful!");
                }
                else if (input.Length == 4 && int.Parse(input) >= int.Parse(currentYear))
                {
                    char yearFirstChar = input[2];
                    char yearSecondChar = input[3];
                    this.year = int.Parse(CharConcatenation.Concat(yearFirstChar, yearSecondChar));
                    Console.WriteLine("SetYear2: " + this.year + " - Successful!");
                    this.yearReady = true;
                }
                else
                {
                    this.yearReady = false;
                    Console.WriteLine("SetYear: " + input + " - Invalid.");
                }
            }
            catch (NullReferenceException)
            {
                Console.WriteLine("SetYear: null");
                year = -1;
                this.yearReady = false;
            }
            Console.WriteLine(yearReady);
        }

        public void SetWriterI(string input)
        {
            if (input == "")
            {
                writerInitial = null;
                Console.WriteLine("SetWriterI: null");
            }
            else if (input == "Type name")
            {
                writerInitial = null;
                Console.WriteLine("SetWriterI: null");
            }
            else
            {
                try
                {
                    string upperInput = input.ToUpper();
                    int firstSpace = upperInput.IndexOf(' ');
                    int secondSpace = upperInput.IndexOf(' ', firstSpace + 1);
                    char firstInitial = upperInput[0];
                    char secondInitial;
                    char thirdInitial;
                    if (firstSpace != -1)
                    {
                        secondInitial = upperInput.Substring(firstSpace + 1)[0];
                        if (secondSpace != -1)
                        {
                            thirdInitial = upperInput.Substring(secondSpace + 1)[0];
                            writerInitial = CharConcatenation.Concat(firstInitial, secondInitial, thirdInitial);
                        }
                        else
                        {
                            writerInitial = CharConcatenation.Concat(firstInitial, secondInitial);
                        }
                    }
                    else
                    {
                        writerInitial = firstInitial.ToString();
                    }
                    Console.WriteLine("SetWriterI: " + writerInitial + " - Successful!");
                }
                catch (NullReferenceException)
                {
                    writerInitial = null;
                    Console.WriteLine("SetWriterI: null");
                }
            }
        }

        public void SetDuration(int input)
        {
            if (input == -1)
            {
                duration = -1;
                durationReady = false;
                Console.WriteLine("SetDuration: null");
            }
            else
            {
                try
                {
                    int checkInput = Array.IndexOf(durations, input);
                    duration = durations[checkInput];
                    durationReady = true;
                    Console.WriteLine("SetDuration: " + duration + " - Successful!");
                }
                catch (IndexOutOfRangeException)
                {
                    durationReady = false;
                    Console.WriteLine("SetDuration: " + input + " - Invalid.");
                }
            }
        }

        public void SetType(string input)
        {
            if (input == null)
            {
                type = null;
                typeReady = false;
                Console.WriteLine("SetType: null");
            }
            else
            {
                try
                {
                    if (input.Equals("Cross"))
                    {
                        type = types[3];
                        Console.WriteLine("SetType: " + type + " - Successful!");
                    }
                    else if (input.Equals("Sim Cross"))
                    {
                        type = types[4];
                        Console.WriteLine("SetType: " + type + " - Successful!");
                    }
                    else
                    {
                        string correctedInput = input.Replace("Sim", " ").Trim();
                        Console.WriteLine("SetTypePre: " + correctedInput + "| SetTypePost: " + correctedInput[0]);
                        int checkInput = Array.IndexOf(types, correctedInput[0].ToString());
                        type = types[checkInput];
                        typeReady = true;
                        Console.WriteLine("SetType: " + type + " - Successful!");
                    }

                }
                catch (IndexOutOfRangeException)
                {
                    typeReady = false;
                    Console.WriteLine("SetType: " + input + " - Invalid.");
                }
            }
        }

        public void SetClientI(string input)
        {
            if (input == "")
            {
                clientInitial = null;
                Console.WriteLine("SetClientI: null");
            }
            else if (input == "Type client name")
            {
                clientInitial = null;
                Console.WriteLine("SetClientI: null");
            }
            else
            {
                try
                {
                    string upperInput = input.ToUpper();
                    int firstSpace = upperInput.IndexOf(' ');
                    int secondSpace = upperInput.IndexOf(' ', firstSpace + 1);
                    int thirdSpace = upperInput.IndexOf(' ', secondSpace + 1);
                    char firstInitial = upperInput[0];
                    char secondInitial;
                    char thirdInitial;
                    char fourthInitial;

                    if (firstSpace != -1)
                    {
                        secondInitial = upperInput.Substring(firstSpace + 1)[0];
                        if (secondSpace != -1)
                        {
                            thirdInitial = upperInput.Substring(secondSpace + 1)[0];
                            if (thirdSpace != -1)
                            {
                                fourthInitial = upperInput.Substring(thirdSpace + 1)[0];
                                clientInitial = CharConcatenation.Concat(firstInitial, secondInitial, thirdInitial, fourthInitial);
                            }
                            else
                            {
                                clientInitial = CharConcatenation.Concat(firstInitial, secondInitial, thirdInitial);
                            }
                        }
                        else
                        {
                            clientInitial = CharConcatenation.Concat(firstInitial, secondInitial);
                        }
                    }
                    else
                    {
                        clientInitial = firstInitial.ToString();
                    }
                    /*
                    if (!input.Contains(" "))
                    {
                        if (input.Length >= 2)
                        {
                            char clientFirstInitial = upperInput[0];
                            char clientSecondInitial = upperInput[1];
                            clientInitial = CharConcatenation.Concat(clientFirstInitial, clientSecondInitial);
                        }
                        else
                        {
                            clientInitial = upperInput[0].ToString();
                        }
                    }
                    else
                    {
                        int secondInitialIndex = input.ToUpper().IndexOf(" ") + 1;
                        char clientFirstInitial = input.ToUpper()[0];
                        char clientSecondInitial = input.ToUpper()[secondInitialIndex];

                        clientInitial = CharConcatenation.Concat(clientFirstInitial, clientSecondInitial);
                    }
                    */
                    Console.WriteLine("SetClientI: " + clientInitial + " - Successful!");
                }
                catch (NullReferenceException)
                {
                    clientInitial = null;
                    Console.WriteLine("SetClientI: null");
                }
            }
         }

        public string Get(string input)
        {
            if (input != null)
            {
                return input;
            }
            else
            {
                return "NULL";
            }
        }
        public string Get(int input)
        {
            if (input != -1)
            {
                return Convert.ToString(input);
            }
            else
            {
                return "NULL";
            }
        }
        public string Get(char input)
        {
            if (input != '\0')
            {
                return Convert.ToString(input);
            }
            else
            {
                return "NULL";
            }
        }

        public string Generate()
        {
            Console.WriteLine("Key Number:");
            keyNumber = Get(market) + Get(year) + Get(writerInitial) + '-' + Get(clientInitial) + '-' + Get(duration) + Get(type) + '-' + number;
            Console.WriteLine(keyNumber);
            number++;
            Properties.Settings.Default.number = number;
            Properties.Settings.Default.Save();

            System.Windows.Forms.Clipboard.SetText(keyNumber);

            return keyNumber;
        }

        public void Save(string property)
        {
            if (property == "market")
            {
                Properties.Settings.Default.market = market;
                Properties.Settings.Default.Save();
            }
            else if (property == "year")
            {
                Properties.Settings.Default.year = year.ToString();
                Properties.Settings.Default.Save();
            }
            else if (property == "writerInitial")
            {
                Properties.Settings.Default.writerInitial = writerInitial.ToString();
                Properties.Settings.Default.Save();
            }
            else if (property == "duration")
            {
                Properties.Settings.Default.duration = duration;
                Properties.Settings.Default.Save();
            }
            else if (property == "type")
            {
                Properties.Settings.Default.type = type;
                Properties.Settings.Default.Save();
            }
            else if (property == "clientInitial")
            {
                Properties.Settings.Default.clientInitial = clientInitial;
                Properties.Settings.Default.Save();
            }
            else if (property == "number")
            {
                Properties.Settings.Default.number = number;
                Properties.Settings.Default.Save();
            }
        }

        public string Load(string property)
        {
            if (property == "market")
            {
                return Properties.Settings.Default.market;
            }
            else if (property == "year")
            {
                return Properties.Settings.Default.year;
            }
            else if (property == "writerInitial")
            {
                return Properties.Settings.Default.writerInitial;
            }
            else if (property == "duration")
            {
                return Properties.Settings.Default.duration.ToString();
            }
            else if (property == "type")
            {
                return Properties.Settings.Default.type;
            }
            else if (property == "clientInitial")
            {
                return Properties.Settings.Default.clientInitial;
            }
            else if (property == "number")
            {
                return Properties.Settings.Default.number.ToString();
            }
            else
            {
                return "NULL";
            }
        }

    }

    static class CharConcatenation
    {
        public static String Concat(params char[] chars)
        {
            if (chars.Length == 0)
            {
                return "";
            }
            StringBuilder s = new StringBuilder(chars.Length);
            foreach (char c in chars)
            {
                s.Append(c);
            }
            return s.ToString();
        }
    }
}
